### PR TITLE
Google Cloud Run deployment script

### PR DESCRIPTION
**Pull request by chrismaree**
This PR adds a script to enable google Cloud Run containers to be deployed and creates an associated cloud scheduler. This contains the following steps:
1) Pull down the config file for the bot from the GCP storage bucket.
2) Deploy the Cloud Run instance with the config file
3) Fetch the deployed Cloud Run URL for the new instance.
4) Computes the default service account using the GCP project number
5) Creates a scheduler job with the associated URL and service account to run the Cloud Run instance daily @ 8AM UTC.

### CODE REVIEW

File: scripts/bot-deployment/DeployReporterCR.sh [hunk 1]
```diff
@@ -0,0 +1,53 @@
+#!/bin/bash
+set -e
+
+if [ $# -ne 1 ]; then
+    echo "Incorrect number of arguments supplied! First and only argument is bot's name. From this the bot's config."
```
Comment (nicholaspai) [inline scripts/bot-deployment/DeployReporterCR.sh:5]: "From this the bot's config" is there meant to be more to this sentence?

File: scripts/bot-deployment/DeployReporterCR.sh [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    echo "example: ./DeployBotCR.sh ethbtc-mainnet-reporter"
```
Comment (nicholaspai) [inline scripts/bot-deployment/DeployReporterCR.sh:6]: Can you include the command, either in an `echo` or in comments, to display all GCR instances available (i.e. which configs can be pulled down)?

File: scripts/bot-deployment/DeployReporterCR.sh (hunk 1 above)
Comment (chrismaree) [reply scripts/bot-deployment/DeployReporterCR.sh]: fixed.

File: scripts/bot-deployment/DeployReporterCR.sh (hunk 2 above)
Comment (chrismaree) [reply scripts/bot-deployment/DeployReporterCR.sh:6]: This pulls the config from the cloud bucket as described in the `Bot Configs` playbook document. I added a comment to (via an `echo`) to explain how to list these.

Comment (chrismaree) [reply scripts/bot-deployment/DeployReporterCR.sh:6]: I also added this to the other deployment script for consistency

File: scripts/bot-deployment/DeployBotGCP.sh [hunk 3]
```diff
@@ -3,7 +3,8 @@ set -e
 
 if [ $# -ne 1 ]; then
     echo "Incorrect number of arguments supplied! First and only argument is bot's name. From this the bot's config and service account will be inferred."
-    echo "example: ./DeployBotGCP.sh ethbtc-monitor-bot"
+    echo "example: ./DeployBotGCP.sh ethbtc-mainnet-monitor"
```
Comment (chrismaree) [inline scripts/bot-deployment/DeployBotGCP.sh:6]: change made to be consistent with @nicholaspai request

File: scripts/bot-deployment/DeployReporterCR.sh (hunk 2 above)
Comment (chrismaree) [reply scripts/bot-deployment/DeployReporterCR.sh:6]: I also added this to the other deployment script for consistency

File: scripts/bot-deployment/DeployBotGCP.sh (hunk 3 above)
Comment (nicholaspai) [reply scripts/bot-deployment/DeployBotGCP.sh:6]: Awesome thanks! nit: you mispelled "available" in line below

File: scripts/bot-deployment/DeployReporterCR.sh (hunk 2 above)
Comment (nicholaspai) [reply scripts/bot-deployment/DeployReporterCR.sh:6]: Thanks! You mispelled "available"

Comment (chrismaree) [reply scripts/bot-deployment/DeployReporterCR.sh:6]: @nicholaspai is the ultimate spell checker.

File: scripts/bot-deployment/DeployBotGCP.sh (hunk 3 above)
Comment (chrismaree) [reply scripts/bot-deployment/DeployBotGCP.sh:6]: fixed

File: scripts/bot-deployment/DeployReporterCR.sh (hunk 2 above)
Comment (nicholaspai) [reply scripts/bot-deployment/DeployReporterCR.sh:6]: LOL

File: scripts/bot-deployment/DeployBotGCP.sh (hunk 3 above)
Comment (nicholaspai) [reply scripts/bot-deployment/DeployBotGCP.sh:6]: looks good

File: scripts/bot-deployment/DeployReporterCR.sh [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    --platform managed \
+    --region=us-central1
+echo "🎉 Bot deployed!"
+
+echo "🧭 Fetching cloud run instance URL"
+
+# Fetch the cloud run URL created for the service.
+cloudRunURL=$(gcloud run services describe $1 --platform managed --region us-central1 --format 'value(status.url)')
+
+echo "🛣  Cloud run URL has been pulled as" $cloudRunURL
```
Comment (mrice32) [inline scripts/bot-deployment/DeployReporterCR.sh:36]: +1 this is very cool.

File: scripts/bot-deployment/DeployReporterCR.sh [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+# Note2: this command assumes that the GCP account account only has 1 project associated with it.
+serviceAccountEmail=$(gcloud projects list --filter="$PROJECT" --format="value(PROJECT_NUMBER)")"-compute@developer.gserviceaccount.com"
+
+echo "📄 Using default service account for project @" $serviceAccountEmail
+
+echo "⏱  Creating cloud scedular to run the cloud run instance"
+
+# Lastly, creat the scheduler job.
+gcloud scheduler jobs create http $1 \
+    --schedule="0 8 * * *" \
```
Comment (mrice32) [inline scripts/bot-deployment/DeployReporterCR.sh:48]: Don't we want 8AM EST, not 8AM UTC? So shouldn't this be `12`, not `8`?

File: scripts/bot-deployment/DeployReporterCR.sh [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+# Copy config files from GCP to the temp file.
+gsutil cp gs://bot-configs/$1.yml $tempFile
+
+echo "✍️  Config has been pulled and placed in a temp directory" $tempFile
+
+echo "🚀 Deploying cloud run instance to GCP"
+
+# Deploy The bot to GCP using the config file.
+gcloud beta run services replace $tempFile \
```
Comment (mrice32) [inline scripts/bot-deployment/DeployReporterCR.sh:26]: Two questions:
1. Does this replace work if there's no existing job?
2. Where does the bot name get fed into this command? It looks like it's referenced in the `services describe` command below? How do those end up matching up?

File: scripts/bot-deployment/DeployReporterCR.sh [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+# Note1: the default service account email is always <project_number>--compute@developer.gserviceaccount.com
+# Note2: this command assumes that the GCP account account only has 1 project associated with it.
+serviceAccountEmail=$(gcloud projects list --filter="$PROJECT" --format="value(PROJECT_NUMBER)")"-compute@developer.gserviceaccount.com"
+
+echo "📄 Using default service account for project @" $serviceAccountEmail
+
+echo "⏱  Creating cloud scedular to run the cloud run instance"
+
+# Lastly, creat the scheduler job.
+gcloud scheduler jobs create http $1 \
```
Comment (mrice32) [inline scripts/bot-deployment/DeployReporterCR.sh:47]: Does this automatically replace the last scheduled job with the same name (to avoid duplicates after running this script multiple times)?

File: scripts/bot-deployment/DeployReporterCR.sh (hunk 5 above)
Comment (chrismaree) [reply scripts/bot-deployment/DeployReporterCR.sh:48]: aha. I see. I can update that and the comment in the bot.

File: scripts/bot-deployment/DeployReporterCR.sh (hunk 6 above)
Comment (chrismaree) [reply scripts/bot-deployment/DeployReporterCR.sh:26]: 1. Yup. Replace is used if you are deploying a new job or upgrading an existing one with Cloud run. Strange syntax, I know.
2. The name and all other information is fed in from the config file. This has the relevant structure as follows:
```
...
metadata:
  name: ethbtc-mainnet-reporter
...
```
This name here matches to the name of the bot in CLoud run.

You do raise a good point, though. It is possible to name the bot differently to the config file. I'll look into a better way of either validating these match, injecting the name into the config or specifying the name of the bot as a param.

File: scripts/bot-deployment/DeployReporterCR.sh (hunk 5 above)
Comment (chrismaree) [reply scripts/bot-deployment/DeployReporterCR.sh]: fixed :)

File: scripts/bot-deployment/DeployReporterCR.sh (hunk 7 above)
Comment (chrismaree) [reply scripts/bot-deployment/DeployReporterCR.sh:47]: If there is a previous scheduler with the same name this command will simply error out, leaving the previous job in place. As a result, you can update the bot Cloud run deployment without needing to update the scheduler.

File: scripts/bot-deployment/DeployReporterCR.sh (hunk 6 above)
Comment (chrismaree) [reply scripts/bot-deployment/DeployReporterCR.sh:26]: I looked a bit further into this and it's not super trivial to use this command in the way I had hoped to validate that the name provided matches the config name. Because we use yaml config files to deploy the cloud run instances (used to store all env variables and cloud run configs) you cant override the `SERVICE` param from a CLI flag.

I've been using this config from the docs [here](https://cloud.google.com/run/docs/deploying#service). From the API [docs](https://cloud.google.com/sdk/gcloud/reference/beta/run/services/replace) we are limited to using only config files to define the `SERVICE` name.

However, I don't really think this is a big deal. As long as the `name` field within the config matches to the actual name of the config file there is no way the names can not line up because the script pulls down the config based on the file name. 

In fact, we have the same restriction with other bot deployment scripts with the `BOT_IDENTIFIER`. There is no validation to check that the name of the bot equals the `BOT_IDENTIFIER`. 

In future we could consider creating a script to validate these configs but for now this would be totally over-engineered imho.

TLDR: we need to always ensure the config names match to the name of the bot in the config itself.

File: scripts/bot-deployment/DeployReporterCR.sh (hunk 7 above)
Comment (chrismaree) [reply scripts/bot-deployment/DeployReporterCR.sh:47]: ye this is a great idea. I've created an [issue](https://github.com/UMAprotocol/protocol/issues/1597) for this and will get on it next.

**Comment by chrismaree** [comment]
> Question: This will not update the docker container correct? The docker container URL I'm assuming is contained in the `bot-configs .yml` file?

Yup, that's right. the docker URL is in the config.
